external/anomaly/ote/task.py

```diff
@@ -131,6 +131,9 @@ def train(
         model_monitor = ModelMonitorCallback(output_model, self.save_model)
         callbacks = [progress, model_monitor]
 
+        if hasattr(self.model, "callbacks"):
+            callbacks.append(self.model.callbacks)
+
```

Comment:
Is this really necessary? I think PL should automatically append the model specific callbacks to the callbacks passed to the trainer.

Reply:
This is just to make it similar to the original task in anomalib. This is in the train.py 
![image](https://user-images.githubusercontent.com/17232914/139877919-83e30f29-47e9-4179-bfbd-4cf806f93727.png)

Reply:
Then maybe both should be changed. This check seems unnecessary

Reply:
@ashwinvaidya17 noticed that early stopping doesnt work as expected. If this is the case, that means that model callbacks are not detected, and this check could be needed. In stfpm early stopping callback is added [here](https://gitlab-icv.inn.intel.com/algo_rnd_team/anomaly/-/blob/development/anomalib/models/stfpm/model.py#L238).

if model callbacks are not appended to `callbacks` list, then the trainer will only take into account `callbacks = [progress, model_monitor]`, which explains why early stopping doesnt work for stfpm

Reply:
According to [the documentation](https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html) the model-specific callbacks should extend the list of callbacks passed to the trainer. I'm guessing this only works if the model-specific callbacks are added by the `configure_callbacks` method of the PL module. We're not using this method to add our model-specific callbacks which is probably why it didn't work earlier. I think it would be better to use the `configure_callbacks` method in the way it was designed so that we keep the task implementation as clean as possible.

Reply:
This is great! Yes, our Lightning implementations do not implement `configure_callbacks` method, but instead assign `self.callbacks`.  Seems like the above implementation then re-invents `configure_callbacks`. Maybe we could prioritise this and update anomalib algos

Reply:
We'll need to make a separate PR in the anomalib repo that updates the PL modules. I'll create a ticket for this. Maybe we can keep the check here for now and remove it once the PL modules are updated.

Reply:
yes agree. i've already created this https://jira.devtools.intel.com/browse/CVS-69866

Reply:
perfect, thanks